File: marketplace-frontend/src/app/main/content/_service/configuration/class-configuration.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Marketplace } from 'app/main/content/_model/marketplace';
import { ClassConfiguration } from '../../_model/meta/configurations';

@Injectable({
  providedIn: 'root'
})
export class ClassConfigurationService {

  constructor(
    private http: HttpClient
  ) { }

  getAllClassConfigurations(marketplace: Marketplace) {
    return this.http.get(`${marketplace.url}/class-configuration/all`);
  }

  getAllClassConfigurationById(marketplace: Marketplace, id: string) {
    return this.http.get(`${marketplace.url}/class-configuration/${id}`);
  }

  getClassConfigurationByName(marketplace: Marketplace, name: string) {
    return this.http.get(`${marketplace.url}/class-configuration/by-name/${name}`);
  }

  getClassConfigurationsByTenantId(marketplace: Marketplace, tenantId: string) {
    return this.http.get(`${marketplace.url}/class-configuration/all/tenant/${tenantId}`);

  }

  getAllForClassConfigurationInOne(marketplace: Marketplace, id: string) {
    return this.http.get(`${marketplace.url}/class-configuration/all-in-one/${id}`);
  }

  createNewEmptyClassConfiguration(marketplace: Marketplace, name: string, description: string) {
    return this.http.post(`${marketplace.url}/class-configuration/new-empty`, [name, description]);
  }

  createNewClassConfiguration(marketplace: Marketplace, tenantId: string, name: string, description: string) {
    const params: string[] = [];
    params.push(tenantId);
    params.push(name);
    params.push(description);

    return this.http.post(`${marketplace.url}/class-configuration/new`, params);
  }

  saveClassConfiguration(marketplace: Marketplace, classConfiguration: ClassConfiguration) {
    return this.http.put(`${marketplace.url}/class-configuration/save`, classConfiguration);
  }


  saveClassConfigurationMeta(marketplace: Marketplace, id: string, name: string, description: string) {
    return this.http.put(`${marketplace.url}/class-configuration/${id}/save-meta/`, [name, description]);
  }

  deleteClassConfiguration(marketplace: Marketplace, id: string) {
    return this.http.delete(`${marketplace.url}/class-configuration/${id}/delete`);
  }

  deleteClassConfigurations(marketplace: Marketplace, ids: string[]) {
    return this.http.put(`${marketplace.url}/class-configuration/delete-multiple`, ids);
  }



}

File: marketplace/marketplace-core/src/main/java/at/jku/cis/iVolunteer/marketplace/configurations/clazz/ClassConfigurationService.java
package at.jku.cis.iVolunteer.marketplace.configurations.clazz;

import java.util.ArrayList;
import java.util.Date;
import java.util.List;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.RequestBody;

import at.jku.cis.iVolunteer.marketplace.configurations.matching.collector.MatchingEntityMappingConfigurationRepository;
import at.jku.cis.iVolunteer.marketplace.meta.core.class_.ClassDefinitionRepository;
import at.jku.cis.iVolunteer.marketplace.meta.core.class_.CollectionService;
import at.jku.cis.iVolunteer.marketplace.meta.core.relationship.RelationshipRepository;
import at.jku.cis.iVolunteer.model.configurations.clazz.ClassConfiguration;
import at.jku.cis.iVolunteer.model.configurations.matching.collector.MatchingEntityMappingConfiguration;
import at.jku.cis.iVolunteer.model.matching.MatchingEntityMappings;
import at.jku.cis.iVolunteer.model.meta.core.clazz.ClassDefinition;
import at.jku.cis.iVolunteer.model.meta.core.relationship.Relationship;

@Service
public class ClassConfigurationService {
	@Autowired
	private ClassConfigurationRepository classConfigurationRepository;
	@Autowired
	private CollectionService collectionService;
	@Autowired
	private MatchingEntityMappingConfigurationRepository matchingCollectorConfigurationRepository;
	@Autowired
	private ClassDefinitionRepository classDefinitionRepository;
	@Autowired
	private RelationshipRepository relationshipRepository;

	@GetMapping("class-configuration/all/tenant/{tenantId}")
	public List<ClassConfiguration> getClassConfigurationsByTenantId(@PathVariable("tenantId") String tenantId) {
		return classConfigurationRepository.findByTenantId(tenantId);
	}

	@GetMapping("class-configuration/{id}")
	public ClassConfiguration getClassConfigurationById(@PathVariable("id") String id) {
		return classConfigurationRepository.findOne(id);
	}

	public ClassConfiguration saveClassConfiguration(@RequestBody ClassConfiguration updatedClassConfiguration) {

		updatedClassConfiguration.setTimestamp(new Date());

		ClassConfiguration classConfiguration = classConfigurationRepository.save(updatedClassConfiguration);

		List<ClassDefinition> classDefinitions = new ArrayList<>();
		classDefinitionRepository.findAll(updatedClassConfiguration.getClassDefinitionIds())
				.forEach(classDefinitions::add);

		if (classDefinitions != null) {
			classDefinitions = updateClassDefinitions(classDefinitions, classConfiguration);
		}

		classDefinitionRepository.save(classDefinitions);

		// Build MatchingCollector
		MatchingEntityMappings mappings = collectionService
				.collectAllClassDefinitionsWithPropertiesAsMatchingEntityMappings(classConfiguration.getId());

		MatchingEntityMappingConfiguration matchingCollectorConfiguration = new MatchingEntityMappingConfiguration();
		matchingCollectorConfiguration.setId(classConfiguration.getId());
		matchingCollectorConfiguration.setClassConfigurationId(classConfiguration.getId());
		matchingCollectorConfiguration.setMappings(mappings);

		matchingCollectorConfigurationRepository.save(matchingCollectorConfiguration);
		return classConfiguration;
	}

	private List<ClassDefinition> updateClassDefinitions(List<ClassDefinition> classDefinitions,
			ClassConfiguration classConfiguration) {

		for (ClassDefinition cd : classDefinitions) {
			cd.setConfigurationId(classConfiguration.getId());
		}

		List<Relationship> relationships = new ArrayList<>();
		relationshipRepository.findAll(classConfiguration.getRelationshipIds()).forEach(relationships::add);

		collectionService.assignLevelsToClassDefinitions(classDefinitions, relationships);

		return classDefinitions;
	}

	public void deleteClassConfiguration(@PathVariable("id") String id) {
		ClassConfiguration classConfiguration = classConfigurationRepository.findOne(id);

		classConfiguration.getClassDefinitionIds().forEach(classDefinitionRepository::delete);
		classConfiguration.getRelationshipIds().forEach(relationshipRepository::delete);

		classConfigurationRepository.delete(id);
	}

	public List<ClassConfiguration> deleteMultipleClassConfigurations(@RequestBody List<String> ids) {
		ids.forEach(this::deleteClassConfiguration);
		return this.classConfigurationRepository.findAll();
	}
	
}
